### ISSUE TITLE
Backtracking failure matching regex `^a?(a?)b?c\1$` against string `abca`

### ISSUE THREAD

**Issue by Davidebyzero** [mrabarnett/mrab-regex#494]
A short program demonstrating this bug:
```
import re, regex
print(str(bool(regex.search(r"^a?(a?)b?c\1$", "abca"))))
```
That program is demonstrated here: [Attempt This Online!](https://ato.pxeger.com/run?1=m72soLIkIz9vwYbM3IL8ohKFolQdIE5PrVhaWpKma3HTpqAoM69EIzUvxba4pEgjKT8_RwMsr1ecmliUnKFRpBSXaK-RaK-ZZJ8cY6iipKOglJiUnKikCQQQMxZAKSgNAA "Python - Attempt This Online")

The output should be `True`, with the regex finding a match, but instead it prints `False`. Apparently the regex engine only tries matching `a?` with `a`, and never backtracks to try an empty match for `a?`.

When `re.search` is used instead of `regex.search` , the bug does not happen. Also changing `a?` to `a??` or changing `b?` to `b??` prevents the bug from happening.

The bug was also present in an old version of mrab-regex: [Try it online!](https://tio.run/##K6gsycjPM/6fmVuQX1SiUJSqA8TpqRX/C4oy80o0UvNSbItLijSS8vNzNMASesWpiUXJGRpFSnGJ9hqJ9ppJ9skxhipKOgpKiUnJiUqaQPD/PwA "Python 3 – Try It Online")

I came across this bug while testing a regex to solve [Visible Dice Faces](https://codegolf.stackexchange.com/questions/149890/visible-dice-faces/259172#259172) with a 41 byte regex using lookbehind, but as it turned out the bug has nothing to do with lookbehind.

**Comment by mrabarnett** [comment]
It should be fixed in regex 2023.03.22, which is being processed by Github Actions as I write this.

**Comment by Davidebyzero** [comment]
Thanks very much for fixing this so quickly!
